polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -111,51 +144,51 @@ class PolyKernel private[kernel] (
                         val newDetail = Option(detail).getOrElse(taskInfo.detail)
                         statusUpdates.publish1(UpdatedTasks(taskInfo.copy(detail = newDetail, progress = (progress * 255).toByte) :: Nil)).unsafeRunSync()
                     }
-                    kernel.runCode(
+
+                    symbolTable.drain() *> interp.runCode(
                       id,
-                      findAvailableSymbols(prevCellIds, kernel),
+                      findAvailableSymbols(prevCellIds, interp),
                       prevCellIds,
                       cell.content.toString
                     ).map {
-                      results => results.evalMap {
-                        case WrapSymbol(symbol) =>
-                          symbolTable.publishAll(symbolTable.RuntimeValue.fromSymbolDecl(symbol).toList).as(None)
-                        case WrapResult(result) =>
-                          IO.pure(Some(result))
-                      }.unNone
+                      results => predefResults merge results.evalMap {
+                        case v: ResultValue =>
+                          symbolTable.publishAll(List(symbolTable.RuntimeValue.fromResultValue(v, interp))).as(v)
+                        case result =>
+                          IO.pure(result)
+                      }
                     }.handleErrorWith {
                       case errs@CompileErrors(_) => IO.pure(Stream.emit(errs))
                       case err@RuntimeError(_) => IO.pure(Stream.emit(err))
                       case err => IO.pure(Stream.emit(RuntimeError(err)))
                     }.guarantee(oq.enqueue1(None)) <* symbolTable.drain()
                 }
-            }.flatten.handleErrorWith(err => IO.pure(Stream.emit(RuntimeError(err))))
+            }.handleErrorWith(err => IO.pure(Stream.emit(RuntimeError(err))))
           }
         Stream.emit(ClearResults()) ++ Stream(cellResults.flatten, oq.dequeue.unNoneTerminate).parJoinUnbounded
     }
   }
 
-  def completionsAt(id: String, pos: Int): IO[List[Completion]] = withKernel(id) {
-    (notebook, cell, kernel) =>
+  def runCells(ids: List[String]): IO[Stream[IO, CellResult]] = getNotebook().map {
+    notebook => Stream.emits(ids).evalMap {
+      id => runCell(id).map(results => results.map(result => CellResult(notebook.path, id, result)))
+    }.flatten // TODO: better control execution order of tasks, and use parJoinUnbounded instead
```

Comment:
can you elaborate on the TODO? If we used `parJoinUnbounded` wouldn't it run the cells in parallel?

Reply:
I think ideally we'd be able to run the outer `IO` for each cell before evaluating any streams, and the streams themselves would handle serializing the execution of the cells. For example, if you click "Run all" I'd expect that all the cells show up in the tasks list as queued (they currently don't because the next cell's `IO` doesn't evaluate until the previous one's `Stream` completes). And they'd run in parallel, because if a cell finishes execution but for some reason still has some "late" messages to publish later and so doesn't end its stream, that shouldn't block the next cell from running.